### ISSUE TITLE
XML root  Name not consistent

### ISSUE THREAD

**Issue by JimWeier-VeteransUnited** [cloudevents/spec#1007]
Why is Root element of XML version named "event",  The C# Implementation its named CloudEvent, , Proto is named CloudEvent, AVRo its named CloudEvent, To be consistent, and to allow inoperability It would be better if the Root Node of XML version were name the same

**Comment by jskeet** [comment]
Well, it's already in a CloudEvents-specific namespace, so personally I think it's unambiguous.

I don't really understand the interoperability argument - @JimWeier-VeteransUnited can you be more specific about how using `<event>` as the element name hurts interop?

**Comment by duglin** [comment]
I don't have a strong opinion about this one. Before @JimWeier-VeteransUnited pointed out the inconsistency I preferred just "event" since the namespace was sufficient to differentiate it from other "event" elements. But, I do like consistency....

**Comment by jskeet** [comment]
> if you an XML with event as the root, and Attempt to Serialize it into your C# Class thats defined as "CloudeEvent", It will fail

If you're using plain XML serialization, I'd *expect* you to have to use attributes to specify the names involved. But fundamentally I wouldn't expect to use XML serialization for this... I'd use the CloudEvents SDK when it has XML format support, which it will when I've finished it. (I've got a prototype already.) You certainly shouldn't serialize a `CloudNative.CloudEvents.CloudEvent` object with a "normal" XML serialization and expect that to follow the XML format.

> I thought the Whole idea behind "CloudEvents" was to standardize the event.... you shouldn't have one standard for XML another for Json Another C# another for some other File Format

C# is *not* a CloudEvent format, whereas XML and JSON are. (Or rather, there are separate CloudEvent formats for JSON and XML, amongs others.)

There's no such thing as "a C# CloudEvent" any more than there's a "Java CloudEvent" as a file format. There's C# code to create, parse, manipulate etc CloudEvents, but that's not the same thing at all.

The purpose of the CloudEvents project isn't to have a *single* standard file format - it's a single standard *data model* which can be represented in multiple (also standardized) formats.

>  If the Format supports Namespaces, they should be the same across the board, If the Format support "Root nodes" they should be the same across the Board .

No - C# namespaces and XML namespaces are very different. They serve similar purposes, but you shouldn't expect to use the URI specified for an XML namespace as a C# namespace, for example.

Basically, I still don't think there's an interop problem *at all* at the moment - and I still prefer `<event>` over `<cloudevent>` as the root element name.

**Comment by JimWeier-VeteransUnited** [comment]
If "The purpose of the CloudEvents project isn't to have a single standard file format - it's a single standard data model which can be represented in multiple (also standardized) formats."
  
 --- If this is the case then event and Cloudevent are NOT the same Data model  you entity may have similar fields/properties but are different data models because there ENTITY name is different, therefore are not a standard Data model, If AVRO is CloudEvent, PROTO is Cloud Event, Then the XML should be CloudEvent as well as JSON

**Comment by JimWeier-VeteransUnited** [comment]
Really CloudEvent is just an Abstract Data Model?   My Bad I will Ignore it completely because Abstract Data Model  I was treating it as a Concrete Data Type.  I will create my own, Its too bad Abstract data Models really dont help with Standards or inoperability between systems if its only abstract

**Comment by jskeet** [comment]
We may be using terminology slightly differently, but I regard it as an abstract data model that is represented concretely in both SDKs and CloudEvent formats. Knowing *what can be represented* in a CloudEvent (e.g. the type system for attributes) and *how CloudEvents can be represented* (in different formats) feels extremely useful to me in terms of standardization.

**Comment by JimWeier-VeteransUnited** [comment]
My point is your Concrete implementations are not Consistent and will cause issues try Something as simple as this (Notice the Class Name is different then the Element that is Created

Create a class
/// <remarks/>
[System.CodeDom.Compiler.GeneratedCodeAttribute("xsd", "4.8.3928.0")]
[System.SerializableAttribute()]
[System.Diagnostics.DebuggerStepThroughAttribute()]
[System.ComponentModel.DesignerCategoryAttribute("code")]
[System.Xml.Serialization.XmlTypeAttribute(AnonymousType = true, Namespace = "http://VU.Common.Schemas.CloudEvent")]
[System.Xml.Serialization.XmlRootAttribute(ElementName ="event", Namespace = "http://VU.Common.Schemas.CloudEvent", IsNullable = false)]
public class CloudEvent
{
 [XmlElement(ElementName = "Id", Namespace = "http://VU.Common.Schemas.CloudEvent", Form = System.Xml.Schema.XmlSchemaForm.Qualified)]
    public string id 
    {
        get
        {
            return _CloudEvent.Id;
        }
        set 
        {
            _CloudEvent.Id = value; 
        }
    }
}

Use a standrd .net XML  Serializer and Serilize it to a file, the using Same standard Serilizer , deserialize it...  it will Fail Then as a developer you need to Dig into the Documentation or Worse yet look at the class to Figure out why, At that point most developers will say abandon it

**Comment by jskeet** [comment]
> Use a standrd .net XML Serializer and Serilize it to a file, the using Same standard Serilizer , deserialize it... it will Fail

Could you clarify why it would fail here? If it does, that sounds like an error in the serializer, given that it's configured to use `<event>` as the element name. If the serializer is so inflexible that it can't cope with anything other than the class name being used as the element name without failing to round-trip, that's not a terribly useful serializer.

But fundamentally, this isn't a use case I'd expect to support, and *in particular* I don't think the class name choices in SDKs should be tied either to each other or to the XML representation. Different file formats will naturally have different conventions - I would be strongly against requiring them to use the exact same names.

**Comment by JimWeier-VeteransUnited** [comment]
Serilization will work fine,  Coming back into the class "De-Serialization" will fail because the XML Root Node Name must match the class Name you trying to De-Serialize  this is out of the Box .net  Functionality...You can override this Functionality but again digging around to figure what its should be,     Stating "Different file formats will naturally have different conventions" then why even have a specification, Just call SpecVersion SV and Data DatNodeElemnt and Time TimeOfEvent and Source Source_Of_Data, If Interoperability and Standard coneventions mean nothing throw it all away... why even have a standard?  Dont put XML in your Cloud Events Standard!  Put it an XMLCloudeventStandard and start over.. 

Stanadardizing CloudEvents is in a way of creating a "Canonical" and ALL Formats whether its JSON Or XML, SOAP,  C#, They should be the same. The purpose of a CDM is to enable the ability to create and distribute a common definition of its ### _**entire** **data** **unit**._ 
This allows for smoother integration between systems, which can help: Improve processes and practices. Make data analytics easier...

**Comment by jskeet** [comment]
I think at this point our views are *so* divergent that there's not much benefit in discussing them further. If you see no value in the CloudEvent specifications as they are at the moment, I suggest you just don't use them. Others (including myself) clearly *do* see value in them.

**Comment by JimWeier-VeteransUnited** [comment]
Fair enough  Are view are different, I believe if you look CloudEvent from a single point of view such as JSON it has value or as Just from XML, or C#  it has Value. But inconsistency across languages and formats just causes unneeded pain points... An working with systems that integrate many many systems, Some using Json, Some XML , Some C# and Rest Some in SOAP Some WCF, Those inconstancies and the effort to try to keep them straight isnt worth the effort... Much Simpler to Take work you have done, Spin off our own and make them consistant.

**Comment by jskeet** [comment]
Just for kicks, I decided to test the claim about XmlSerializer - it worked fine for me. Complete code:

Program.cs:
```csharp
using System.Xml.Serialization;

namespace XmlSerializationTest;

class Program
{
    static void Main()
    {
        var originalEvent = new CloudEvent { Id = "test-id" };
        var serializer = new XmlSerializer(typeof(CloudEvent));

        var writer = new StringWriter();
        serializer.Serialize(writer, originalEvent);

        string xml = writer.ToString();
        Console.WriteLine($"XML: {xml}");
        var parsedEvent = (CloudEvent) serializer.Deserialize(new StringReader(xml));
        Console.WriteLine($"Parsed ID: {parsedEvent.Id}");
    }
}
```

CloudEvent.cs:

```csharp
using System.Xml.Serialization;

namespace XmlSerializationTest;

[XmlRoot(ElementName = "event", Namespace = "http://cloudevents.io/xmlformat/V1")]
public class CloudEvent
{
    [XmlElement(ElementName = "id", Namespace = "http://cloudevents.io/xmlformat/V1")]
    public string Id { get; set; }
}
```

Result:

```text
XML: <?xml version="1.0" encoding="utf-16"?><event xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns="http://cloudevents.io/xmlformat/V1"><id>test-id</id></event>
Parsed ID: test-id
```

No problem...

**Comment by JimWeier-VeteransUnited** [comment]
Now take that Serialized XML Pull back in and DE-SERIALIZE it into a CloudEvent

**Comment by jskeet** [comment]
> Now take that Serialized XML Pull back in and DE-SERIALIZE it into a CloudEvent

Um, that's exactly what I did:

```csharp
 var parsedEvent = (CloudEvent) serializer.Deserialize(new StringReader(xml));
```

**Comment by JimWeier-VeteransUnited** [comment]
Sorry Missed that

**Comment by jskeet** [comment]
> Much Simpler to Take work you have done, Spin off our own and make them consistant.

If you only want to deal with a single format, then go ahead - just be aware that you'll end up with a much more constrained system, whereas embracing the model of "CloudEvents have *this* data model, which can be expressed in *these* different ways" (and with multiple different languages etc) gives much more flexibility. If you're fine without that flexibility, that's no problem - but many other contributors *do* want that flexibility, otherwise we wouldn't have come up with multiple event formats...

**Comment by jskeet** [comment]
> What ever platform or format the DataEntity Name will be CloudEvent there is absolutely no value in deviating from it.

Except that you'll then be self-consistent, and inconsistent with *every other platform supporting CloudEvents*. Basically it will be your own variation of CloudEvents. That's your choice of course, and it may even make sense if you have multiple platforms to support internally, but never need to interface with any external systems.

(It's still unclear to me where you'd expect the name "CloudEvent" to come into a JSON document, mind you - unless you're proposing adding a wrapper object for the sole purpose of introducing the name. That seems somewhat wasteful and pointless to me though.)

**Comment by JimWeier-VeteransUnited** [comment]
Why would my Inconsistent naming my Cloudevent break other System and keep me from integratingwith them?   I should be able to do it just fine, You being inconsistent across platforms will work for you, But not for me?

CLOUD Event Come into play in JSON when you turn on a feature that enables  you make root value (usually JSON Object but can be any type) wrapped within a single property JSON object, where key as the "root name", as determined by annotation introspector (esp. for JAXB that uses @XmlRootElement.name) or fallback (non-qualified class name). Feature is mostly intended for JAXB compatibility.

**Comment by jskeet** [comment]
> Why would my Inconsistent naming my Cloudevent break other System and keep me from integratingwith them?

Because if the CloudEvents XML format specifies a root element name of `<event>` and you decided to use `<CloudEvent>` instead, that's simply not a valid CloudEvent representation in terms of deviating from the specification. I don't think that should come as a surprise.

**Comment by duglin** [comment]
Lively discussion! :-)
I just want to make sure I understand the state of things... it seems to me that @JimWeier-VeteransUnited was concerned about two aspects of the XML format:
1 - technical perspective:  inconsistency between the XML naming and the code naming ("event" in XML but "CloudEvent" in the code/classes). However, I think @jskeet showed that from a coding perspective this shouldn't be an issue since XML (de)serializers should support custom name-mappings between what's on the wire and what the code expects. @JimWeier-VeteransUnited, it wasn't clear to me if you're claiming that some languages don't support this level of customizations.
2 - mental perspective: inconsistency across the various CE formats (e.g. "event" in XML but "CloudEvent" in one of the other format or the SDKs). As stated in 1 it's not about whether the code can work with different words but rather is there a mental mapping that devs might need to do because of the inconsistency. As an extreme example, what if xml used `<dog>` instead of `<event>` - perfectly legal and the code can work, but are we making life harder for humans? 

Do I have that right?

If so, I think @JimWeier-VeteransUnited is asking a valid question about the naming inconsistency - and it might really just come down to a personal preference type of decision. But I agree with @jskeet that the value of CloudEvents isn't lessened due to the current choice since from a coding perspective is seems to work just fine.

**Comment by JimWeier-VeteransUnited** [comment]
Yes Lively discussion .... I love it ! I like this because  together as a community we will get the best product delivered. No matter what you all decide... but At minimum it has brought forth thought... I thank you all for that...

And Yes your over view his correct

**Comment by duglin** [comment]
We has a quick chat about this on today's call and there wasn't a big desire to change the element name from "event" to "CloudEvent" so the group decided to keep it as is for now.
